presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -317,7 +317,7 @@ private void processNewSources()
                 }
 
                 // if current operator is finished...
-                if (current.isFinished()) {
+                if (current.isFinished() && !next.isFinished() && isBlocked(next).isDone()) {
```

Comment:
@dain note also that `isBlocked(next).isDone()` is not the same as `operator.isBlocked().isDone()`. Difference lies in memory allocation (operator that is not _internally_ blocked can still be blocked from Driver's perspective if waits for memory). Calling `finish` in such state sounds like not-a-problem unless Operator allocates even more memory in finish (as `HashBuilderOperator` does).